File: C/min_stack.c
typedef struct {
    int* stack;
    int size;
    int min;
} MinStack;

/** initialize your data structure here. */

MinStack* minStackCreate() {
    MinStack* minstack = (MinStack*)malloc(sizeof(MinStack));
    
    minstack->stack = malloc(10000 * sizeof(int));
    minstack->size = 0;
    minstack->min = INT_MAX;
    
    return minstack;
}

void minStackPush(MinStack* obj, int x) {
    obj->stack[obj->size] = x;
    obj->size ++;
    if(x < obj->min) obj->min = x;
}

void minStackPop(MinStack* obj) {
    -- obj->size;
    if(obj->size == 0) obj->min = INT_MAX; 
    else {
        obj->min = obj->stack[0];
        for(int i = 0; i < obj->size; i++) {
            if(obj->stack[i] < obj->min) obj->min = obj->stack[i];
        }
    }
}

int minStackTop(MinStack* obj) {
    return obj->stack[obj->size-1];
}

int minStackGetMin(MinStack* obj) {
    return obj->min;
}

void minStackFree(MinStack* obj) {
    
}

/**
 * Your MinStack struct will be instantiated and called as such:
 * MinStack* obj = minStackCreate();
 * minStackPush(obj, x);
 
 * minStackPop(obj);
 
 * int param_3 = minStackTop(obj);
 
 * int param_4 = minStackGetMin(obj);
 
 * minStackFree(obj);
*/

File: Python3/min_stack.py
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack = []
        self.min = None

    def push(self, x: int) -> None:
        self.stack.append(x)
        if self.min is None or x < self.min: self.min = x

    def pop(self) -> None:
        self.stack.pop()
        if self.min not in self.stack:
            self.min = min(self.stack) if len(self.stack) > 0 else None

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
